session/filetransfer.go

```diff
@@ -0,0 +1,580 @@
+		case h.msgChan <- msg:
+		}
+
+	case wsft.MessageTypeError:
+		select {
+		// If there's an active async handler, pass the error down,
+		// otherwise, log the error.
+		case h.mutex <- struct{}{}:
+			<-h.mutex
+			var erro wsft.Error
+			err := msgpack.Unmarshal(msg.Body, &erro)
+			if err != nil {
+				log.Errorf("Error decoding error message from client: %s", err.Error())
+			} else {
+				log.Errorf("Received error from client: %s", *erro.Error)
+			}
+		case h.msgChan <- msg:
+		}
+
+	default:
+		h.Error(msg, w, errors.Errorf(
+			"session: filetransfer message type '%s' not supported",
+			msg.Header.MsgType,
+		))
+	}
+}
+
+func (h *FileTransferHandler) StatFile(msg *ws.ProtoMsg, w ResponseWriter) {
+	var params StatFile
+	err := msgpack.Unmarshal(msg.Body, &params)
+	if err != nil {
+		h.Error(msg, w, errors.Wrap(err, "malformed request parameters"))
+		return
+	} else if err = params.Validate(); err != nil {
+		h.Error(msg, w, errors.Wrap(err, "invalid request parameters"))
+		return
+	}
+	stat, err := os.Stat(*params.Path)
+	if err != nil {
+		h.Error(msg, w, errors.Wrapf(err,
+			"failed to get file info from path '%s'", *params.Path))
+		return
+	}
+	mode := uint32(stat.Mode())
+	size := stat.Size()
+	modTime := stat.ModTime()
+	fileInfo := wsft.FileInfo{
+		Path:    params.Path,
+		Size:    &size,
+		Mode:    &mode,
+		ModTime: &modTime,
+	}
+	if statT, ok := stat.Sys().(*syscall.Stat_t); ok {
+		// Only return UID/GID if the filesystem/OS supports it
+		fileInfo.UID = &statT.Uid
+		fileInfo.GID = &statT.Gid
+	}
+	b, _ := msgpack.Marshal(fileInfo)
+
+	err = w.WriteProtoMsg(&ws.ProtoMsg{
+		Header: ws.ProtoHdr{
+			Proto:     ws.ProtoTypeFileTransfer,
+			MsgType:   wsft.MessageTypeFileInfo,
+			SessionID: msg.Header.SessionID,
+		},
+		Body: b,
+	})
+	if err != nil {
+		log.Errorf("error sending FileInfo to client: %s", err.Error())
+	}
+}
+
+// chunkWriter is used for packaging writes into ProtoMsg chunks before
+// sending it on the connection.
+type chunkWriter struct {
+	SessionID string
+	Offset    int64
+	W         ResponseWriter
+}
+
+func (c *chunkWriter) Write(b []byte) (int, error) {
+	msg := ws.ProtoMsg{
+		Header: ws.ProtoHdr{
+			Proto:     ws.ProtoTypeFileTransfer,
+			MsgType:   wsft.MessageTypeChunk,
+			SessionID: c.SessionID,
+			Properties: map[string]interface{}{
+				"offset": c.Offset,
+			},
+		},
+		Body: b,
+	}
+	err := c.W.WriteProtoMsg(&msg)
+	if err != nil {
+		return 0, err
+	}
+	c.Offset += int64(len(b))
+	return len(b), err
+}
+
+func (h *FileTransferHandler) InitFileDownload(msg *ws.ProtoMsg, w ResponseWriter) (err error) {
+	params := new(GetFile)
+	defer func() {
+		if err != nil {
+			log.Error(err.Error())
+			h.Error(msg, w, err)
+		}
+	}()
+	if err := msgpack.Unmarshal(msg.Body, params); err != nil {
+		err = errors.Wrap(err, "malformed request parameters")
+		return err
+	} else if err = params.Validate(); err != nil {
+		err = errors.Wrap(err, "invalid request parameters")
+		return err
+	}
```

Comment:
I'm not 100% sure, but I believe you are not writing the "err" variable returned to the function inside the "if" for error handling of msgpack.Unmarshal, because you are initializing a new variable (:=).

Reply:
You're right. However, I think that the `return err` assigns the new variable to the old one. I changed it just to be sure.